domino/datasets.py

```diff
@@ -0,0 +1,198 @@
+            target_chunk_size: int,
+
+    ):
+        self.csrf_no_check_header = csrf_no_check_header
+        self.dataset_id = dataset_id
+        self.local_path_file_or_directory = local_path_to_file_or_directory
+        self.log = log
+        self.request_manager = request_manager
+        self.routes = routes
+        self.target_chunk_size = target_chunk_size or 8 * MB
+        self.file_upload_setting = file_upload_setting or FILE_UPLOAD_SETTING_DEFAULT
+        self.max_workers = max_workers or MAX_WORKERS
+        self.target_relative_path = target_relative_path
+        self.upload_key = None  # this will be set once the session is started
+
+    def __enter__(self):
+        # creating upload session
+        start_upload_body = {
+            "filePaths": [],
+            "fileCollisionSetting": self.file_upload_setting
+        }
+        start_upload_url = self.routes.datasets_start_upload(self.dataset_id)
+        self.upload_key = self.request_manager.post(start_upload_url, json=start_upload_body).json()
+        if not self.upload_key:
+            raise RuntimeError(f"upload key for {self.dataset_id} not found. Session could not start.")
+        # starting uploads
+        try:
+            paths = self._upload()
+            return paths
+        except Exception as e:
+            self.log.error("Upload failed. See error for details.")
+            raise e
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        # catching errors
+        if not self.upload_key:
+            raise RuntimeError(f"upload key for {self.dataset_id} not found. Could not end session.")
+        if exc_type != None:
+            self.log.error(f"Upload for dataset {self.dataset_id} and file {self.local_path_file_or_directory} failed, "
+                           f"attempting to cancel session. Please try again.")
+            self.log.error(f"Error type: {exc_val}. Error message: {exc_tb}.")
+            if not isinstance(exc_type, ValueError):
+                self._cancel_upload_session() # is it is a ValueError, canceling session would fail
+            return
+        # ending snapshot uplaod
+        try:
+            url = self.routes.datasets_end_upload(self.dataset_id, self.upload_key, self.target_relative_path)
+            self.request_manager.get(url)
+            self.log.info("Upload session completed successfully.")
+        except Exception as e:
+            self.log.error("Ending snapshot upload failed. See error for details. Attempting to cancel "
+                           "upload session.")
+            self._cancel_upload_session()
+
+    def _cancel_upload_session(self):
+        url = self.routes.datasets_cancel_upload(self.dataset_id, self.upload_key)
+        try:
+            self.request_manager.get(url)
+            self.log.info("Upload session cancelled successfully.")
+        except Exception as e:
+            self.log.info("Cancelling upload session failed. See error for details.")
+            raise e
```

Comment:
probably want to prevent the exception from being re-raise (see previous comment)